core/templates/components/question-directives/question-player/services/question-player-state.service.ts

```diff
@@ -17,79 +17,86 @@
-    };
+  private _hintUsed(question): void {
+    var questionId = question.getId();
+    if (!this.questionPlayerState[questionId]) {
+      this.createNewQuestionPlayerState(
+        questionId,
+        question.getLinkedSkillIds());
+    }
+    this.questionPlayerState[questionId].usedHints.push(
+      {timestamp: this.getCurrentTime()});
+  }
 
-    var _answerSubmitted = function(
-        question, isCorrect, taggedSkillMisconceptionId) {
-      var questionId = question.getId();
-      if (!questionPlayerState[questionId]) {
-        createNewQuestionPlayerState(
-          questionId,
-          question.getLinkedSkillIds());
-      }
-      // Don't store a correct answer in the case where
-      // the learner viewed the solution for this question.
-      if (isCorrect && questionPlayerState[questionId].viewedSolution) {
-        return;
-      }
-      questionPlayerState[questionId].answers.push(
-        {isCorrect: isCorrect,
-          timestamp: getCurrentTime(),
-          taggedSkillMisconceptionId: taggedSkillMisconceptionId
-        });
-    };
+  private _solutionViewed(question): void {
+    var questionId = question.getId();
+    if (!this.questionPlayerState[questionId]) {
+      this.createNewQuestionPlayerState(
+        questionId,
+        question.getLinkedSkillIds());
+    }
+    this.questionPlayerState[questionId].viewedSolution = {
+      timestamp: this.getCurrentTime()};
+  }
 
+  private _answerSubmitted(
+      question, isCorrect, taggedSkillMisconceptionId): void {
+    var questionId = question.getId();
+    if (!this.questionPlayerState[questionId]) {
+      this.createNewQuestionPlayerState(
+        questionId,
+        question.getLinkedSkillIds());
+    }
+    // Don't store a correct answer in the case where
+    // the learner viewed the solution for this question.
+    if (isCorrect && this.questionPlayerState[questionId].viewedSolution) {
+      return;
+    }
+    this.questionPlayerState[questionId].answers.push(
+      {isCorrect: isCorrect,
+        timestamp: this.getCurrentTime(),
+        taggedSkillMisconceptionId: taggedSkillMisconceptionId
+      });
+  }
 
-    return {
-      hintUsed: function(question) {
-        _hintUsed(question);
-      },
-      solutionViewed: function(question) {
-        _solutionViewed(question);
-      },
-      answerSubmitted: function(
-          question, isCorrect, taggedSkillMisconceptionId) {
-        _answerSubmitted(question, isCorrect, taggedSkillMisconceptionId);
-      },
-      getQuestionPlayerStateData: function() {
-        return questionPlayerState;
-      }
-    };
-  }]);
+
+  hintUsed(question): void {
+    this._hintUsed(question);
+  }
+  solutionViewed(question): void {
+    this._solutionViewed(question);
+  }
+  answerSubmitted(
+      question, isCorrect, taggedSkillMisconceptionId): void {
+    this._answerSubmitted(question, isCorrect, taggedSkillMisconceptionId);
+  }
+  getQuestionPlayerStateData() {
```

Comment:
Missing return type. Ditto here and everywhere.

Reply:
Done, but StateObjectFactory and QuestionObjectFactory don't have predefined interfaces so I'm returning 'Object' currently. Should I implement Interfaces for these?